server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +462,274 @@ public int hashCode() {
-                    // If the flow comes here, then we should have a open shard available on node.
-                    if (currentKeysToClean.contains(
-                        new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId)
-                    )) {
+                    // Update staleKeysCount based on specific readerCacheKeyId, then remove it from the countMap
+                    countMap.computeIfPresent(cleanupKey.readerCacheKeyId, (k, v) -> {
+                        staleKeysCount.addAndGet(v);
+                        // Return null to remove the key after updating staleKeysCount
+                        return null;
+                    });
+
+                    // Check if countMap is empty after removal to decide if we need to remove the shardId entry
+                    if (countMap.isEmpty()) {
+                        return null; // Returning null removes the entry for shardId
+                    }
+                }
+                return countMap; // Return the modified countMap to keep the mapping
+            });
+        }
+
+        /**
+         * Clean cache based on stalenessThreshold
+         */
+        void cleanCache() {
+            cleanCache(stalenessThreshold);
+        }
+
+        /**
+         * Force Clean cache without checking stalenessThreshold
+         */
+        private void forceCleanCache() {
+            cleanCache(0);
+        }
+
+        /**
+         * Cleans the cache based on the provided staleness threshold.
+         * <p>If the percentage of stale keys in the cache is less than this threshold,the cache cleanup process is skipped.
+         * @param threshold The staleness threshold as a double.
+         */
+        private void cleanCache(double threshold) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Cleaning Indices Request Cache with threshold : " + threshold);
+            }
+            writeLock.lock();
+            try {
+                if (canSkipCacheCleanup(threshold)) {
+                    return;
+                }
+                // Contains CleanupKey objects with open shard but invalidated readerCacheKeyId.
+                final Set<CleanupKey> cleanupKeysFromOutdatedReaders = new HashSet<>();
+                // Contains CleanupKey objects of a closed shard.
+                final Set<Object> cleanupKeysFromClosedShards = new HashSet<>();
+
+                processCleanupKeys(cleanupKeysFromOutdatedReaders, cleanupKeysFromClosedShards);
+
+                if (cleanupKeysFromOutdatedReaders.isEmpty() && cleanupKeysFromClosedShards.isEmpty()) {
+                    return;
+                }
+
+                for (Iterator<Key> iterator = cache.keys().iterator(); iterator.hasNext();) {
+                    Key key = iterator.next();
+                    if (shouldRemoveKey(key, cleanupKeysFromOutdatedReaders, cleanupKeysFromClosedShards)) {
                         iterator.remove();
```

Comment:
this iterator on the cache keys to perform remove is not supported in the Ehcache based implementation and it will throw UnsupportedException. Ref [here](https://github.com/opensearch-project/OpenSearch/blob/main/plugins/cache-ehcache/src/main/java/org/opensearch/cache/store/disk/EhcacheDiskCache.java#L390). So curious how the cleanup will happen in that case ? Also is the cleanup logic protected by any update to the cache since iterator will be changing the cache state?

It seems like the IT tests for cleanup logic with EhCache is missing currently, we should add it for both Ehcache and TieredCache.

Reply:
Great catch, spoke with @sgup432 , he will create a fix in a different PR to fix the underlying issue in 
[EhcacheDiskCache.java](https://github.com/opensearch-project/OpenSearch/blob/57f6f46bb6d1b1dba0e89cf4ad3b3c4ef0abcb12/plugins/cache-ehcache/src/main/java/org/opensearch/cache/store/disk/EhcacheDiskCache.java#L390)

Reply:
@sohami Good catch! I will add remove() on existing EhCacheKeyIterator to address this as part of separate PR.
So we can keep using iterator.remove() here.

Regarding IT, I will add more for TieredCache. For Ehcache, currently serialization changes are still in review which is blocking IT for ehcache. I already had a plan to add once we are unblocked.

Reply:
Linking Sagar's PR here for better visibility https://github.com/opensearch-project/OpenSearch/pull/12653